core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java

```diff
@@ -575,9 +575,11 @@ assert rowTypesAreEquivalent(
           }
         }
         final MutableRel next = MutableRels.preOrderTraverseNext(queryDescendant);
+        final int currentInputIndex = queryDescendant.getInputs().size() > globalInputIndex
+            ? globalInputIndex : 0;
         final MutableRel childOrNext =
             queryDescendant.getInputs().isEmpty()
-                ? next : queryDescendant.getInputs().get(0);
+                ? next : queryDescendant.getInputs().get(currentInputIndex);
```

Comment:
I think getInputs.get(currentInputIndex) is just a walk-around to force traverse the missed nodes. Actually, preOrderTraverseNext method can traverse all the nodes of this plan tree, There is no need to add a new status to record the current node's position.

Reply:
It is a demo.

I think that always `get(0)`  is error. 
We should find a better way to traverse the missing nodes.

Reply:
I think get(0) is correct, get(0) is just a starting point when traverse child-nodes, preOrderTraverseNext method can traverse the missing nodes.